core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -625,12 +642,28 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      * @return <tt>true</tt> if this settings object contains no settings
      */
     public boolean isEmpty() {
-        return this.settings.isEmpty(); // TODO: account for secure settings
+        return this.settings.isEmpty() && (secureSettings == null || secureSettings.getSettingNames().isEmpty());
     }
 
     /** Returns the number of settings in this settings object. */
     public int size() {
-        return this.settings.size(); // TODO: account for secure settings
+        return keySet().size();
+    }
+
+    /** Returns the fully qualified setting names contained in this settings object. */
+    public Set<String> keySet() {
```

Comment:
Not sure what you mean by a copy? It lazily constructs it the keyset, but there is only one copy.